In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

##Install BeautifulSoup and Requests

In [3]:
!pip install BeautifulSoup
!pip install Requests

  Using cached https://files.pythonhosted.org/packages/1e/ee/295988deca1a5a7accd783d0dfe14524867e31abb05b6c0eeceee49c759d/BeautifulSoup-3.2.1.tar.gz
    ERROR: Complete output from command python setup.py egg_info:
    ERROR: Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-install-s60ulxme/BeautifulSoup/setup.py", line 22
        print "Unit tests have failed!"
                                      ^
    SyntaxError: Missing parentheses in call to 'print'. Did you mean print("Unit tests have failed!")?
    ----------------------------------------
ERROR: Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-install-s60ulxme/BeautifulSoup/


In [4]:
!pip install BeautifulSoup4

In [5]:
pip install lxml

Note: you may need to restart the kernel to use updated packages.


## Import Beautiful Soup

In [1]:
from urllib.request import urlopen
import requests
from bs4 import BeautifulSoup

print('Done')

Done


In [2]:
url = requests.get("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M").text

### Map URL and interpret HTML Table

In [3]:
soup = BeautifulSoup(url,'lxml')
type(soup)
#print(soup.prettify())

bs4.BeautifulSoup

## Identify Postal Code Table

In [4]:
My_table = soup.find('table',{'class':'wikitable sortable'})
#My_table

In [5]:
import csv

## Fetch Content from HTML Table into Rows and Columns

In [28]:
list_of_rows= []   
rows  = My_table.find_all('tr')

for row in rows:
    list_of_cells = []
    for cell in row.find_all('td'):
        list_of_cells.append(cell.text)
    list_of_rows.append(list_of_cells)
list_of_rows

[[],
 ['M1A', 'Not assigned', 'Not assigned\n'],
 ['M2A', 'Not assigned', 'Not assigned\n'],
 ['M3A', 'North York', 'Parkwoods\n'],
 ['M4A', 'North York', 'Victoria Village\n'],
 ['M5A', 'Downtown Toronto', 'Harbourfront\n'],
 ['M5A', 'Downtown Toronto', 'Regent Park\n'],
 ['M6A', 'North York', 'Lawrence Heights\n'],
 ['M6A', 'North York', 'Lawrence Manor\n'],
 ['M7A', "Queen's Park", 'Not assigned\n'],
 ['M8A', 'Not assigned', 'Not assigned\n'],
 ['M9A', 'Etobicoke', 'Islington Avenue\n'],
 ['M1B', 'Scarborough', 'Rouge\n'],
 ['M1B', 'Scarborough', 'Malvern\n'],
 ['M2B', 'Not assigned', 'Not assigned\n'],
 ['M3B', 'North York', 'Don Mills North\n'],
 ['M4B', 'East York', 'Woodbine Gardens\n'],
 ['M4B', 'East York', 'Parkview Hill\n'],
 ['M5B', 'Downtown Toronto', 'Ryerson\n'],
 ['M5B', 'Downtown Toronto', 'Garden District\n'],
 ['M6B', 'North York', 'Glencairn\n'],
 ['M7B', 'Not assigned', 'Not assigned\n'],
 ['M8B', 'Not assigned', 'Not assigned\n'],
 ['M9B', 'Etobicoke', 'Cloverdale

In [30]:
outfile=open("./postcodes.csv","w")
writer =csv.writer(outfile)
writer.writerow(["Postal Code","Borough","Neighbourhood"])
writer.writerows(list_of_rows)

### Generate CSV file and migrate the HTML List into CSV file

In [31]:
#import necessary modules

with open('./postcodes.csv','rt')as f:
  data = csv.reader(f)
  for row in data:
        print(row)

['Postal Code', 'Borough', 'Neighbourhood']
[]
['M1A', 'Not assigned', 'Not assigned\n']
['M2A', 'Not assigned', 'Not assigned\n']
['M3A', 'North York', 'Parkwoods\n']
['M4A', 'North York', 'Victoria Village\n']
['M5A', 'Downtown Toronto', 'Harbourfront\n']
['M5A', 'Downtown Toronto', 'Regent Park\n']
['M6A', 'North York', 'Lawrence Heights\n']
['M6A', 'North York', 'Lawrence Manor\n']
['M7A', "Queen's Park", 'Not assigned\n']
['M8A', 'Not assigned', 'Not assigned\n']
['M9A', 'Etobicoke', 'Islington Avenue\n']
['M1B', 'Scarborough', 'Rouge\n']
['M1B', 'Scarborough', 'Malvern\n']
['M2B', 'Not assigned', 'Not assigned\n']
['M3B', 'North York', 'Don Mills North\n']
['M4B', 'East York', 'Woodbine Gardens\n']
['M4B', 'East York', 'Parkview Hill\n']
['M5B', 'Downtown Toronto', 'Ryerson\n']
['M5B', 'Downtown Toronto', 'Garden District\n']
['M6B', 'North York', 'Glencairn\n']
['M7B', 'Not assigned', 'Not assigned\n']
['M8B', 'Not assigned', 'Not assigned\n']
['M9B', 'Etobicoke', 'Cloverdale\n'

### Read CSV File into Data frame

In [32]:
import pandas as pd
result=pandas.read_csv('./postcodes.csv')
df = pd.DataFrame(result)
df

Postal Code           Borough            Neighbourhood
0           M1A      Not assigned           Not assigned\n
1           M2A      Not assigned           Not assigned\n
2           M3A        North York              Parkwoods\n
3           M4A        North York       Victoria Village\n
4           M5A  Downtown Toronto           Harbourfront\n
..          ...               ...                      ...
283         M8Z         Etobicoke              Mimico NW\n
284         M8Z         Etobicoke     The Queensway West\n
285         M8Z         Etobicoke  Royal York South West\n
286         M8Z         Etobicoke         South of Bloor\n
287         M9Z      Not assigned           Not assigned\n

[288 rows x 3 columns]

## Clean the table by removing EoL

In [33]:
df['Neighbourhood'].replace(r'\s+|\\n', ' ', regex=True, inplace=True)
print(df)

    Postal Code           Borough           Neighbourhood
0           M1A      Not assigned           Not assigned 
1           M2A      Not assigned           Not assigned 
2           M3A        North York              Parkwoods 
3           M4A        North York       Victoria Village 
4           M5A  Downtown Toronto           Harbourfront 
..          ...               ...                     ...
283         M8Z         Etobicoke              Mimico NW 
284         M8Z         Etobicoke     The Queensway West 
285         M8Z         Etobicoke  Royal York South West 
286         M8Z         Etobicoke         South of Bloor 
287         M9Z      Not assigned           Not assigned 

[288 rows x 3 columns]


# Remove Boroughs from data frame where value is Not assigned

In [101]:
df1= df[df['Borough']!= 'Not assigned']
df1

Postal Code           Borough              Neighbourhood
2           M3A        North York                 Parkwoods 
3           M4A        North York          Victoria Village 
4           M5A  Downtown Toronto              Harbourfront 
5           M5A  Downtown Toronto               Regent Park 
6           M6A        North York          Lawrence Heights 
..          ...               ...                        ...
282         M8Z         Etobicoke  Kingsway Park South West 
283         M8Z         Etobicoke                 Mimico NW 
284         M8Z         Etobicoke        The Queensway West 
285         M8Z         Etobicoke     Royal York South West 
286         M8Z         Etobicoke            South of Bloor 

[211 rows x 3 columns]

## 77 rows were not considered where Borough = Not Assigned

In [102]:
df1.sort_values(by=['Postal Code'])

Postal Code      Borough    Neighbourhood
11          M1B  Scarborough           Rouge 
12          M1B  Scarborough         Malvern 
29          M1C  Scarborough      Port Union 
28          M1C  Scarborough      Rouge Hill 
27          M1C  Scarborough  Highland Creek 
..          ...          ...              ...
232         M9V    Etobicoke     Mount Olive 
234         M9V    Etobicoke   South Steeles 
235         M9V    Etobicoke     Thistletown 
233         M9V    Etobicoke     Silverstone 
245         M9W    Etobicoke       Northwest 

[211 rows x 3 columns]

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


Postal Code           Borough              Neighbourhood  Is Duplicated
2           M3A        North York                 Parkwoods           False
3           M4A        North York          Victoria Village           False
4           M5A  Downtown Toronto              Harbourfront           False
5           M5A  Downtown Toronto               Regent Park           False
6           M6A        North York          Lawrence Heights           False
..          ...               ...                        ...            ...
282         M8Z         Etobicoke  Kingsway Park South West           False
283         M8Z         Etobicoke                 Mimico NW           False
284         M8Z         Etobicoke        The Queensway West           False
285         M8Z         Etobicoke     Royal York South West           False
286         M8Z         Etobicoke            South of Bloor           False

[211 rows x 4 columns]

## CHeck rows where Neighbourhood values is Not Assigned

In [103]:
print("Check Not Assigned values in Neighbourhood column:\n") 
df2 = df1['Neighbourhood'].str.contains('Not assigned') 
df2

Check Not Assigned values in Neighbourhood column:



2      False
3      False
4      False
5      False
6      False
       ...  
282    False
283    False
284    False
285    False
286    False
Name: Neighbourhood, Length: 211, dtype: bool

###Results shows there are no rows where Neighbourhood value is not assigned 

## Now merge neighbourhood with same Post codes and Borough names into single row

In [105]:
df2 = (df1.groupby(['Postal Code', 'Borough'])['Neighbourhood'].apply(list))
df2 

Postal Code  Borough    
M1B          Scarborough                                   [Rouge , Malvern ]
M1C          Scarborough          [Highland Creek , Rouge Hill , Port Union ]
M1E          Scarborough               [Guildwood , Morningside , West Hill ]
M1G          Scarborough                                            [Woburn ]
M1H          Scarborough                                         [Cedarbrae ]
                                                  ...                        
M9N          York                                                   [Weston ]
M9P          Etobicoke                                           [Westmount ]
M9R          Etobicoke      [Kingsview Village , Martin Grove Gardens , Ri...
M9V          Etobicoke      [Albion Gardens , Beaumond Heights , Humbergat...
M9W          Etobicoke                                           [Northwest ]
Name: Neighbourhood, Length: 103, dtype: object

## find shape of the cleaned data set

In [106]:
df2.shape

(103,)

## On Clubbing the Neighbourhoods, total data set size is 103 rows

In [107]:
import pandas as pd
coord = pd.read_csv('http://cocl.us/Geospatial_data')
coord

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [120]:
result3 = pd.merge(df2,coord,on='Postal Code',how='left')
result3

Postal Code                                      Neighbourhood   Latitude  \
0           M1B                                 [Rouge , Malvern ]  43.806686   
1           M1C        [Highland Creek , Rouge Hill , Port Union ]  43.784535   
2           M1E             [Guildwood , Morningside , West Hill ]  43.763573   
3           M1G                                          [Woburn ]  43.770992   
4           M1H                                       [Cedarbrae ]  43.773136   
..          ...                                                ...        ...   
98          M9N                                          [Weston ]  43.706876   
99          M9P                                       [Westmount ]  43.696319   
100         M9R  [Kingsview Village , Martin Grove Gardens , Ri...  43.688905   
101         M9V  [Albion Gardens , Beaumond Heights , Humbergat...  43.739416   
102         M9W                                       [Northwest ]  43.706748   

     Longitude  
0   -79.194353  
1   -79.160497  
2   -79.188711  
3   -79.216917  
4   -79.239476  
..         ...  
98  -79.518188  
99  -79.532242  
100 -79.554724  
101 -79.588437  
102 -79.594054  

[103 rows x 4 columns]